pkg/compute/bidder.go

```diff
@@ -165,37 +221,141 @@ func (b Bidder) ReturnBidResult(
 // |      false        |      false         |       N/A         |       N/A          |   false   |   false    |
 func (b Bidder) doBidding(
 	ctx context.Context,
-	request bidstrategy.BidStrategyRequest,
-	calculator capacity.UsageCalculator,
-) (*bidstrategy.BidStrategyResponse, models.Resources, error) {
+	job *models.Job,
+	resourceUsage *models.Resources) (*bidStrategyResponse, error) {
+	// NB(forrest): allways run semantic bidding before resource bidding
 	// Check semantic bidding strategies before calculating resource usage.
-	semanticResponse, err := b.semanticStrategy.ShouldBid(ctx, request)
+	semanticResponse, err := b.runSemanticBidding(ctx, job)
 	if err != nil {
-		return nil, models.Resources{}, fmt.Errorf("error asking bidding strategy if we should bid: %w", err)
+		return nil, fmt.Errorf("semantic bidding: %w", err)
 	}
 
 	// we shouldn't bid, and we're not waiting, bail.
-	if !semanticResponse.ShouldBid && !semanticResponse.ShouldWait {
-		return &semanticResponse, models.Resources{}, nil
+	if !semanticResponse.bid && !semanticResponse.wait {
+		return semanticResponse, nil
 	}
 
-	// the request is semantically biddable or waiting, calculate resource usage and check resource-based bidding.
-	parsedUsage, err := request.Job.Task().ResourcesConfig.ToResources()
+	// else the request is semantically biddable or waiting, calculate resource usage and check resource-based bidding.
+	resourceResponse, err := b.runResourceBidding(ctx, job, resourceUsage)
 	if err != nil {
-		return nil, models.Resources{}, fmt.Errorf("error parsing resources config: %w", err)
+		return nil, fmt.Errorf("resource bidding: %w", err)
 	}
-	resourceUsage, err := calculator.Calculate(ctx, request.Job, *parsedUsage)
-	if err != nil {
-		return nil, models.Resources{}, fmt.Errorf("error calculating resource requirements for job: %w", err)
+
+	return &bidStrategyResponse{
+		bid:                 resourceResponse.bid,
+		wait:                semanticResponse.wait || resourceResponse.wait,
+		reason:              resourceResponse.reason,
+		calculatedResources: resourceResponse.calculatedResources,
+	}, nil
+}
+
+func (b Bidder) runSemanticBidding(
+	ctx context.Context,
+	job *models.Job,
+) (*bidStrategyResponse, error) {
+	// ask the bidding strategy if we should bid on this job
+	request := bidstrategy.BidStrategyRequest{
+		NodeID:   b.nodeID,
+		Job:      *job,
+		Callback: b.getApproveURL(),
+	}
+
+	// assume we are bidding unless a request is rejected
+	shouldBid := true
+	// assume we're not waiting to bid unless a request indicates so.
+	shouldWait := false
+	reasons := make([]string, 0, len(b.semanticStrategy))
+	for _, s := range b.semanticStrategy {
+		// TODO(forrest): this can be parallelized with a wait group, although semantic checks ought to be quick.
```

Comment:
Worth adding a ticket, otherwise it's unlikely anyone will come back to this and measure whether it's worth running these concurrently

Reply:
My thinking with this TODO was that if/when performance becomes an issue here, an engineer will eventually debug the code to this point, see the comment and give it a try. It's meant to be a breadcrumb in-case there is an issue. Alright if we don't create an issue for this?